dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/thread/ObservableBlockingQueue.kt

```diff
@@ -0,0 +1,49 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.thread
+
+import com.datadog.android.internal.thread.NamedRunnable
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicLong
+
+internal open class ObservableBlockingQueue<E : Any>(
+    capacity: Int,
+    private val currentTimeProvider: () -> Long = { System.currentTimeMillis() }
+) : LinkedBlockingQueue<E>(capacity) {
+
+    private var lastDumpTimestamp: AtomicLong = AtomicLong(0)
+
+    fun dumpQueue(): Map<String, Int>? {
+        val currentTime = currentTimeProvider.invoke()
+        val last = lastDumpTimestamp.get()
+        val timeSinceLastDump = currentTime - last
+        return if (timeSinceLastDump > DUMPING_TIME_INTERVAL_IN_MS) {
+            if (lastDumpTimestamp.compareAndSet(last, currentTime)) {
+                buildDumpMap()
+            } else {
+                null
+            }
+        } else {
+            null
+        }
+    }
```

Comment:
maybe this logic should be on consumer side, because it creates some indeterministic behavior from the caller perspective, but I'm okay keeping it like that, because it is only for the internal use.